app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts

```diff
@@ -0,0 +1,220 @@
+  | CreateMetaMetricsDataDeletionTaskAction
+  | UpdateDataDeletionTaskStatusAction
+  | SetHasMetaMetricsDataRecordedAction;
+
+// Type for the messenger of MetaMetricsDataDeletionController
+export type MetaMetricsDataDeletionControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof controllerName,
+    MetaMetricsDataDeletionControllerMessengerActions,
+    never,
+    never,
+    never
+  >;
+
+/**
+ * Controller responsible for maintaining
+ * state related to Metametrics data deletion
+ */
+export class MetaMetricsDataDeletionController extends BaseController<
+  typeof controllerName,
+  MetaMetricsDataDeletionState,
+  MetaMetricsDataDeletionControllerMessenger
+> {
+  #dataDeletionService: PublicInterface<DataDeletionService>;
+
+  #getMetaMetricsId: () => string | null;
+
+  /**
+   * Creates a MetaMetricsDataDeletionController instance.
+   *
+   * @param args - The arguments to this function.
+   * @param args.dataDeletionService - The service used for deleting data.
+   * @param args.messenger - Messenger used to communicate with BaseV2 controller.
+   * @param args.state - Initial state to set on this controller.
+   * @param args.getMetaMetricsId - A function that returns the current MetaMetrics ID.
+   */
+  constructor({
+    dataDeletionService,
+    messenger,
+    state,
+    getMetaMetricsId,
+  }: {
+    dataDeletionService: PublicInterface<DataDeletionService>;
+    messenger: MetaMetricsDataDeletionControllerMessenger;
+    state?: MetaMetricsDataDeletionState;
+    getMetaMetricsId: () => string | null;
+  }) {
+    // Call the constructor of BaseControllerV2
+    super({
+      messenger,
+      metadata,
+      name: controllerName,
+      state: { ...defaultState, ...state },
+    });
+    this.#getMetaMetricsId = getMetaMetricsId;
+    this.#dataDeletionService = dataDeletionService;
+  }
+
+  /**
+   * Creating the delete regulation using source regulation
+   *
+   */
+  async createMetaMetricsDataDeletionTask(): Promise<void> {
+    const metaMetricsId = this.#getMetaMetricsId();
+    if (!metaMetricsId) {
+      throw new Error('MetaMetrics ID not found');
+    }
+
+    const { data } =
+      await this.#dataDeletionService.createDataDeletionRegulationTask(
+        metaMetricsId,
+      );
+    this.update((state) => {
+      state.metaMetricsDataDeletionId = data?.regulateId;
+      state.metaMetricsDataDeletionDate = Date.now();
+      return state;
+    });
+    await this.updateDataDeletionTaskStatus();
+  }
+
+  /**
+   * To check the status of the current delete regulation.
+   */
+  async updateDataDeletionTaskStatus(): Promise<void> {
+    const deleteRegulationId = this.state.metaMetricsDataDeletionId;
+    if (deleteRegulationId.length === 0) {
+      throw new Error('Delete Regulation id not found');
+    }
+
+    const { data } =
+      await this.#dataDeletionService.fetchDeletionRegulationStatus(
+        deleteRegulationId,
+      );
+
+    const regulation = data?.regulation;
```

Comment:
nit: According to the types, it looks like `data` and `regulation` should never be undefined. Is there's a possibility that they're undefined at runtime? If so, do we need to explicitly handle that case?

```suggestion
    const { data: { regulation } } =
      await this.#dataDeletionService.fetchDeletionRegulationStatus(
        deleteRegulationId,
      );

```

Reply:
As long as the api fetch is successful, it should be returning `regulation`